### DISCUSSION TITLE
App Router Error: Rendered more hooks than during the previous render

### DISCUSSION THREAD

**Discussion by matallui** [vercel/next.js#59493] [App Router] [answered]
Hi!

I'm hitting an issue with App Router in which React throws an error about different hook calls during rerender.

I would open a bug instead, but I can't provide a reproduction link since this is a private repo.
I'm also not sure how to create a reproduction example, since I don't fully understand the problem.

Any chance anyone else has seen this? Or can point me in the right direction?
I'm happy to test this locally and potentially open a PR, just trying to get some guidance first.

For context, this seems to happen when I call `router.replace()`. I've also been able to hit the issue in production when calling `router.refresh()`. Somehow seems to be related to calling a `refresh()` at a random "inconvenient" time for Next or something. Either way, seems like it could be a bug in App Router, since I wouldn't expect to see this in any scenario.

This example is running Next 14.0.0 but I've tested with 14.0.4 and got the same error.

Thanks in advance!

<img width="1262" alt="Screenshot 2023-12-11 at 5 18 57 PM" src="https://github.com/vercel/next.js/assets/6828071/b88ad05b-3bb7-45c4-adf5-e84fc2334986">

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Are you calling `router.replace` or any method under the router, while rendering? as in:

```jsx
const App = () => {
  if (condition) router.replace()
 
 return <div />
}
```

That'll lead to disaster, because it is setting state as your render, which means the render is not pure, which means you are potentially batching more state updates, while trying to reflect an state update.

- https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes good read here

The root cause being that, `router` methods (replace, push, etc) are now using React's state management, useReducer. You can install redux-devTools and see the actions being dispatched! (Because the Next.js team connected their useReducer to the redux-devTools) ~ 

Anyway, could you see if this is the issue?

**Reply by matallui** [reply]
Oh, I see... Now that you mention, the culprit call was being made when a firebase auth event was being triggered, so it is likely it was happening while rendering.

I will have that into account and come up with a way to schedule to router call after render. What would be a good way to do that in React? This event callback is actually contained within a React component, so I can probably just make sure to keep things in sync with React.

**Reply by icyJoseph** [reply]
Depending on the situation, use-effect, or as part of a user action

**Reply by gentrith44belt** [reply]
@matallui did you resolve this?

I have kind of similar situation, so only for a specific page (not-found.tsx) when I navigate back to home, it is throwing this error, and only when I access a 404 page directly from url, something like myapp.com/en/wehfugwf

**Reply by gentrith44belt** [reply]
```typescript
const router = useRouter();
const pathname = usePathname();
const [isPending, setTransition] = useTransition();

const setLanguage = (locale: Locale) => {
setTransition(() => {
	router.replace(pathname, { locale });
	router.refresh();
});
};
``` 
     
as @icyJoseph mentioned this router.replace and router.refresh was breaking the app for me

**Reply by matallui** [reply]
Yeah, for me the solution was just to not call `router` methods on an async event handler. I forget exactly what I did, but the issue for me was what I described above.

**Reply by bartoszhernas** [reply]
I have same issue, error happens from NextJS internal. Page is fully server rendered (app router), and I have an `if (!user) { return notFound(); }` which crashes on production. No idea what to do here. I don't have any hooks in this component (as it's server component).

**Reply by bartoszhernas** [reply]
![CleanShot 2025-02-09 at 18 36 26@2x](https://github.com/user-attachments/assets/53585889-72fa-41e6-9bf2-0a4aa6170076)
I was able to reproduce this locally (by using hot reload), but normally happens only in production.

**Reply by Reflex2468** [reply]
@bartoszhernas I'm experiencing the same issue. Locally I've also been able to get it sometimes through HMR but in general it seems very hard to reliably reproduce. Does it always happen to you or is it also only sometimes?

**Reply by bartoszhernas** [reply]
I've created a new issue with that https://github.com/vercel/next.js/issues/75841 

I keep having this issue, seen in production on https://fragplace.com when clicking eg. on a single perfumer

Improvement from version 14 is that main `/loading.tsx` is not appearing

**Reply by pythonicode** [reply]
@bartoszhernas @Reflex2468 I am having the same issue with a server `page.tsx`:

```tsx
export default async function Home() {
    if (process.env.NEXT_PUBLIC_ENVIRONMENT === 'production') {
        return redirect('my-website');
    }

    return null;
}
```

> Rendered more hooks than during the previous render.

I am shocked I can't find anything else about this online? This is a major issue in our app that cannot be resolved by any typical debugging?

**Reply by bartoszhernas** [reply]
It now also happens after clicking on magic link with `next-auth v5 beta25` library

**Reply by heecheon92** [reply]
> @bartoszhernas @Reflex2468 I am having the same issue with a server `page.tsx`:
> 
> ```tsx
> export default async function Home() {
>     if (process.env.NEXT_PUBLIC_ENVIRONMENT === 'production') {
>         return redirect('my-website');
>     }
> 
>     return null;
> }
> ```
> 
> > Rendered more hooks than during the previous render.
> 
> I am shocked I can't find anything else about this online? This is a major issue in our app that cannot be resolved by any typical debugging?

I think this method pattern should be allowed. I mean, converting page.tsx into route.ts kind of resolves the issue but page.tsx should be able to use "redirect" based on a condition.

**Comment by Machinarius** [comment]
I am wondering if this could be triggering for us from a React Server Component? We have an RSC as our main `src/app/page.tsx` that is in charge of `next/navigation#redirect`-ing our users to particular areas of our app after they log-in. Given how that function works by injecting a `<meta>` tag so the Next client-side code performs the actual redirection, I am wondering if that's why we are triggering this error?

We are using Next 15.2.0

**Comment by bartoszhernas** [comment]
Temporary fix is to use client side redirection instead.

Example code: 
``` tsx
  // instead of:
  // redirect(profileUrl(session.user));
  // https://github.com/vercel/next.js/issues/57455
  return <RedirectionPage redirectUrl={profileUrl(session.user)} replace />;
```

``` tsx
import SimpleContainer from '@/containers/components/SimpleContainer';
import { Card, CardBody, CardText, CardTitle } from 'reactstrap';
import React from 'react';
import { LoadingIcon } from '@/containers/components/LoadingIcon';
import RedirectionPageClient from '@/containers/RedirectionPageClient';

export default function RedirectionPage({ redirectUrl, replace }: { redirectUrl?: string; replace?: boolean }) {
  return (
    <>
      {redirectUrl && <RedirectionPageClient redirectUrl={redirectUrl} replace={replace} />}
      <SimpleContainer>
        <Card className="p-4">
          <CardBody className="text-center">
            <CardText className={'mx-auto mb-3 opacity-50'}>
              <LoadingIcon isLg />
            </CardText>
            <CardTitle tag="h2">Redirecting</CardTitle>
            <CardText>Please wait a second</CardText>
          </CardBody>
        </Card>
      </SimpleContainer>
    </>
  );
}
```

``` tsx
export default function RedirectionPageClient({ redirectUrl, replace }: { redirectUrl: string; replace?: boolean }) {
  const router = useRouter();
  useEffect(() => {
    if (replace) {
      router.replace(redirectUrl);
    } else {
      router.push(redirectUrl);
    }
  }, [redirectUrl, replace, router]);
  return null;
}
```

**Reply by EthicalJobsJesse** [reply]
Raising this just in case it helps anyone else. We were running into this issue consistently and it ended up being a race condition with an external script being initialized at the same time as a server side redirect was happening. Not too sure what part of the script was doing it, but I'm guessing it was trying to modify some values of the window history at the same time as NextJS and that caused the error.

For anyone experiencing this issue but having no luck elsewhere, I'd recommend trying to disable or delay <Script /> loading if possible to see if it's a loading/initialization conflict.
